### PR TITLE
Feat/recipe execution

### LINKED ISSUE #190: connect the list of ingredients with the list of foodItems in individualRecipeScreen
In the individual recipe screen, you have a list of ingredients. I show if you currently have them

This will be very hard

### PR DESCRIPTION
**Pull request by ricardo-perello**
# [Feat/recipe execution #301] Add Recipe Execution Flow

---

## **Description**
This PR introduces a comprehensive feature for executing recipes step-by-step, enhancing the user experience and bridging functionality gaps in our application. The following changes have been implemented:

---

### 1. **Recipe Execution State Machine**
- A `RecipeExecutionState` was introduced to manage the sequential navigation between:
  - **Select Servings**: Allows users to adjust the recipe servings dynamically.
  - **Select Food Items**: Enables users to select and allocate food items for each recipe ingredient.
  - **Instructions**: Displays step-by-step cooking instructions with navigation controls.

### 2. **ViewModel Integration**
- Implemented `ExecuteRecipeViewModel` to manage the state and business logic:
  - Handling food item consumption temporarily.
  - Transitioning between states and updating UI dynamically.
  - Maintaining local and Firestore sync for food items.

### 3. **Screens**
- **`ServingsScreen`**:
  - Adjust servings with clear "+" and "-" controls.
  - "Next" button transitions to the food selection screen.
- **`SelectFoodItemsForIngredientScreen`**:
  - Lists available food items for each ingredient.
  - Includes sliders for allocating quantities to each ingredient.
  - Automatically consumes selected items temporarily.
- **`InstructionScreen`**:
  - Displays individual instructions for the recipe.
  - Back and Next buttons for navigation.
  - Finish button appears at the last step.

### 4. **Food Item Management**
- Added the ability to set food items (`setFoodItems`) in the repository to reflect temporary and permanent consumption of food items in Firestore.

### 5. **Logging**
- Added detailed logs throughout the ViewModel and UI for debugging and tracking state transitions, user interactions, and data updates.

---

## **Key Files and Changes**

1. **ViewModel**
   - `ExecuteRecipeViewModel`: Manages state transitions and the entire recipe execution logic.

2. **UI**
   - `RecipeExecutionScreen.kt`: State-based rendering for the execution flow.
   - `ServingsScreen.kt`, `SelectFoodItemsForIngredientScreen.kt`, `InstructionScreen.kt`: Modular implementations for each state.
   - `FoodItemSelectionCard.kt`: Reusable component for managing food item allocation.

3. **Repository**
   - `FoodItemRepository.kt` & `FoodItemRepositoryFirestore.kt`: Enhanced functionality for setting and syncing food items in Firestore.

4. **Navigation**
   - Updated navigation graph to include the **Recipe Execution** flow under its route.

---

## **User Flow**

1. Users start the flow by pressing "Start Recipe" on a recipe card.
2. Adjust servings and proceed to food selection.
3. Allocate food items for each ingredient, consuming them temporarily.
4. Follow instructions step-by-step until completion.
5. Finish button takes users back to the recipe overview.

---

### Manual Tests
- Adjust servings: Validated UI updates and serving persistence.
- Temporary consumption: Ensured food items are decremented locally but synced back correctly in Firestore.
- Instructions navigation: Tested edge cases for "Back," "Next," and "Finish" buttons.

### Automated Tests
- Pending addition.

---

## **Next Steps**
- Add automated tests for ViewModel and UI functionality.
- Gather feedback for UI/UX enhancements.

---

## **Acknowledgments**
A big thanks to the team for their support and feedback in integrating this complex feature seamlessly. Special mention to [@ricardo-perello](https://github.com/ricardo-perello) for leading the effort on this PR.

---

## **Merge Checklist**

- [x] ViewModel logic reviewed and tested.
- [x] UI components integrated with state.
- [x] Repository methods tested for Firestore updates.
- [x] Navigation verified for all states.

---

**Ready for review and merge!** 🚀

### CODE REVIEW

File: app/src/main/java/com/android/shelfLife/ui/recipes/IndividualRecipe/IndividualRecipe.kt [hunk 1]
```diff
@@ -197,9 +221,9 @@ private fun IndividualRecipeScreenPreviewEasterEgg() {
   IndividualRecipeScreen(navigationActions = navigationActions)
 }
 // this preview shows the example where we do have a selected recipe
-// @Preview()
-// @Composable
-// private fun IndividualRecipeScreenPreview() {
+//@Preview()
+//@Composable
```
Comment (Dearop) [inline app/src/main/java/com/android/shelfLife/ui/recipes/IndividualRecipe/IndividualRecipe.kt]: Is the preview really necessary? Why is it commented?

File: app/src/main/java/com/android/shelfLife/ui/recipes/execution/FoodItemSelector.kt [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+
+    Log.d("SelectFoodItemsScreen", "Current ingredient: $ingredientName")
+    Log.d(
+        "SelectFoodItemsScreen",
+        "Available food items: ${availableFoodItems.joinToString { it.foodFacts.name }}"
+    )
+    Log.d(
+        "SelectFoodItemsScreen",
+        "Currently selected items: $currentlySelectedItems"
+    )
```
Comment (Dearop) [inline app/src/main/java/com/android/shelfLife/ui/recipes/execution/FoodItemSelector.kt:63]: great usage of logging to help with debugging!

File: app/src/main/java/com/android/shelfLife/ui/recipes/execution/RecipeExecution.kt [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+import com.android.shelfLife.ui.navigation.NavigationActions
+import com.android.shelfLife.ui.navigation.Route
+import com.android.shelfLife.viewmodel.recipes.RecipeExecutionState
+import com.android.shelfLife.viewmodel.recipes.ExecuteRecipeViewModel
+
+@Composable
+fun RecipeExecutionScreen(navigationActions: NavigationActions, viewModel: ExecuteRecipeViewModel = hiltViewModel()) {
+    val currentState by viewModel.state.collectAsState()
+
+    when (currentState) {
```
Comment (Dearop) [inline app/src/main/java/com/android/shelfLife/ui/recipes/execution/RecipeExecution.kt:24]: great usage of when instead of ifs

File: app/src/main/java/com/android/shelfLife/ui/recipes/execution/FoodItemSelector.kt [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+
+    if (ingredientName == null) {
+        Log.e("SelectFoodItemsScreen", "Ingredient name is null. Navigating back.")
+        navigationActions.goBack()
+        return
+    }
+
+    Scaffold(
+        topBar = {
+            TopAppBar(
```
Comment (Dearop) [inline app/src/main/java/com/android/shelfLife/ui/recipes/execution/FoodItemSelector.kt:73]: Use Custom top app bar

File: app/src/main/java/com/android/shelfLife/viewmodel/recipes/ExecuteRecipeViewModel.kt
Comment (Dearop) [inline app/src/main/java/com/android/shelfLife/viewmodel/recipes/ExecuteRecipeViewModel.kt:1]: Current Issue: MutableStateFlow and StateFlow are used for multiple small pieces of data (e.g., _currentIngredientIndex, _currentInstructionIndex, etc.), which can increase complexity.
* Improvement: Group related states into a single data class.
* This reduces the number of flows and makes the ViewModel state easier to manage.

The implementation would look like this : 
```
data class RecipeExecutionData(
    val currentState: RecipeExecutionState,
    val servings: Float,
    val currentIngredientIndex: Int,
    val currentInstructionIndex: Int,
    val selectedFoodItems: Map<String, List<FoodItem>>,
    val availableFoodItems: List<FoodItem>
)
private val _executionData = MutableStateFlow(RecipeExecutionData(...))
val executionData: StateFlow<RecipeExecutionData> = _executionData.asStateFlow()
```

**Review by Dearop — CHANGES_REQUESTED** [review]
## Pr Reveiew
This PR introduces a comprehensive “Recipe Execution” feature, adding functionality for executing recipes step-by-step.

I've added comments for specific issues, however I also feel like the recipe screens need some animations when adding instructions such as : 
```
AnimatedContent(
            targetState = currentInstruction,
            transitionSpec = {
                fadeIn() with fadeOut() // Define fade animations
            }
        ) { instruction ->
            Text(
                text = instruction,
                style = MaterialTheme.typography.bodyLarge.copy(fontSize = 20.sp),
                modifier = Modifier.padding(16.dp)
            )
        }
```



## Impact

This PR adds a well-structured and user-focused recipe execution feature, aligning with modern app development practices and providing a seamless experience for users. It showcases thoughtful architecture, clean state management, and robust Firestore integration, making it a valuable addition to the app’s functionality.

File: app/src/main/java/com/android/shelfLife/ui/recipes/execution/FoodItemSelector.kt (hunk 4 above)
Comment (ricardo-perello) [reply app/src/main/java/com/android/shelfLife/ui/recipes/execution/FoodItemSelector.kt:73]: the custom bar doesnt allow me to use other icons such as the Icons.Default.Close, which i need.

File: app/src/main/java/com/android/shelfLife/ui/recipes/IndividualRecipe/IndividualRecipe.kt (hunk 1 above)
Comment (ricardo-perello) [reply app/src/main/java/com/android/shelfLife/ui/recipes/IndividualRecipe/IndividualRecipe.kt]: its not, this relates to a previous issue. it shouldve been deleted a long time ago.
